pipeline_steps/gate.groovy

```diff
@@ -0,0 +1,194 @@
+            failed = antecedentTriggerBuilds.findAll() {it.getResult() && it.getResult().isWorseThan(success)}
+            if (failed) {
+                println("Antecedent builds that have completed with failure:\n${failed.join("\n")}")
+                println("Killing existing builds that depend on failed antecedents.")
+                for (gb in gateBuilds) {
+                    gb.doKill()
+                }
+                gateBuilds = []
+            }else if (antecedentTriggerBuilds){
+                completed = antecedentTriggerBuilds.findAll() {!(it in currentAntecedentTriggerBuilds)}
+                println("Antecedent builds that have completed with success:\n${completed.join("\n")}")
+            }
+            antecedentTriggerBuilds = currentAntecedentTriggerBuilds
+        }
+
+        if (antecedentTriggerBuilds.empty) {
+            // A failed gate build only causes the trigger build to fail if all antecedents succeed
+            failedGateBuilds = gateBuilds.findAll() {it.getResult() && it.getResult().isWorseThan(success)}
+            if (failedGateBuilds) {
+                errMsg = "One or more builds failed to reach success:\n${failedGateBuilds.collect() {it.getAbsoluteUrl()}.join("\n")}"
+                println(errMsg)
+                throw new Exception(errMsg)
+            }
+        }
+
+        pullRequestIDs = antecedentTriggerBuilds.collect(){getPullRequestID(it)}
+        pullRequestIDs.add(getPullRequestID(triggerBuild))
+        println("Base branch being tested against: ${ghprbTargetBranch}.")
+        println("Pull requests to merge on top of base branch in tests: ${pullRequestIDs}.")
+        for (jobName in gateJobNames) {
+            if (! gateBuilds.any {it.getParent().getName() == jobName}) {
+                job = Hudson.instance.getItems(org.jenkinsci.plugins.workflow.job.WorkflowJob).find {it.displayName == jobName}
+                nextNumber = job.getNextBuildNumber()
+                println("Requesting new build:")
+                build(
+                    job: jobName,
+                    wait: false,
+                    parameters: [
+                        [
+                            $class: "StringParameterValue",
+                            name: "RPC_GATING_BRANCH",
+                            value: RPC_GATING_BRANCH,
+                        ],
+                        [
+                            $class: "StringParameterValue",
+                            name: "BRANCH",
+                            value: ghprbTargetBranch,
+                        ],
+                        [
+                            $class: "StringParameterValue",
+                            name: "pullRequestChain",
+                            value: common.dumpCSV(pullRequestIDs),
+                        ],
+                    ]
+                )
+                while(true){
```

Comment:
Could you kick off all the builds in a loop, then in a second loop, wait for queuing and find each of the new builds? Then the queue processing (which may involve nodepool requests) can happen in parallel.

Reply:
As we discussed, I've been working on this already. It doesn't make a great deal of difference but for projects with a lot of tests there will be a small improvement. I'm looking to remove the loop completely and have the checking done as part of the outer loop's task.